### PR TITLE
feat: Allow customizing TSDB postings decoder

### PR DESCRIPTION
**Pull request by yeya24**
This pr adds the counterpart of https://github.com/prometheus/prometheus/pull/13242 to specify a custom postings decoder so that TSDB and Compactor is able to decode the postings and read index.

By default, the default `index.DecodePostingsRaw` will be used.

I will keep this pr in Draft first and it would be great to get some early feedback.

### CODE REVIEW

**Comment by yeya24** [comment]
@GiedriusS Can you help review this as well?

File: tsdb/index/index.go
Comment (GiedriusS) [inline tsdb/index/index.go:1857]: Nice :+1:

**Review by GiedriusS — COMMENTED** [review]
I think it looks good, let me try to hook this into Thanos and see if these changes are enough for the Compactor

**Comment by GiedriusS** [comment]
> @GiedriusS Can you help review this as well?
> 
> I think one thing I am not sure is that how to make the hook easier to use if we want to use it in Thanos. I think we need to choose different PostingsDecoder based on block `meta.json` file. However, Prometheus doesn't have that information and it is kind of hard to pass meta.json info in Thanos when we initialize the compactor [here](https://github.com/thanos-io/thanos/blob/main/cmd/thanos/compact.go#L333)

I'm thinking maybe the postings encoder could be encoded into the compaction group's name? :thinking: Of course that would forbid from compacting blocks with different encodings but it's probably a one time switch either way? :thinking: passing down Thanos specific meta-information to the compactor would probably require invasive changes to Prometheus code and I'm not sure if the maintainers would accept that.

File: tsdb/compact.go
Comment (GiedriusS) [inline tsdb/compact.go:481]: Maybe it would be enough to pass `Block` to the decoder function so that we could dynamically use a different postings decoder?

Comment (yeya24) [reply tsdb/compact.go:481]: Then do we get the Thanos metadata using `Block`? I wonder if we can change the `tsdb.BlockMeta` struct of TSDB itself...
Actually https://github.com/prometheus/prometheus/blob/main/tsdb/block.go#L211 might be a place we can extend to contain postings format info without changing Thanos. WDYT

Comment (yeya24) [reply tsdb/compact.go]: Added the code to the latest commit

**Comment by GiedriusS** [comment]
Got a hacked-together version working with the current code: https://github.com/thanos-io/thanos/pull/7109. This means that it should be enough to get this merged to unblock our project

File: tsdb/db.go
Comment (bboreham) [inline tsdb/db.go:685]: Why doesn't this one use the factory?

File: tsdb/db.go
Comment (bboreham) [inline tsdb/db.go:769]: I don't think you should alter the options passed in.

File: tsdb/index/index.go
Comment (bboreham) [inline tsdb/index/index.go:1157]: Why isn't this parameterized?

File: tsdb/head_read.go
Comment (bboreham) [inline tsdb/head_read.go:152]: Unrelated change.

**Review by bboreham — COMMENTED** [review]
Note #12304 adds a symbol-table to the decoder; I don't think that changes much in what you are trying to do but please check if the APIs are compatible.

File: tsdb/head_read.go
Comment (yeya24) [reply tsdb/head_read.go]: Fixed

File: tsdb/index/index.go
Comment (yeya24) [reply tsdb/index/index.go]: Fixed

File: tsdb/db.go
Comment (yeya24) [reply tsdb/db.go]: Removed

**Comment by GiedriusS** [comment]
Could anyone take a look? :disappointed:

File: tsdb/index/index.go
Comment (bboreham) [reply tsdb/index/index.go:1857]: Can you explain what `Decoder` is for?  I don't think I looked at it before, but before this PR it was a wrapper for a function pointer, which was exported.
The only use in Prometheus was in `Reader`, as an unexported data member.

This PR gives us both `Decoder` and `PostingsDecoder` in this file, which also leaves me confused.  There is a `PostingsEncoder` but no `Encoder`.  What's going on?

I can't see where this newly-added member `Postings` (why that name?) is used in this PR.

**Review by bboreham — COMMENTED** [review]
I'm confused.

File: tsdb/index/index.go
Comment (GiedriusS) [reply tsdb/index/index.go:1857]: Previously the function that decoded postings was called Postings() so to make fewer changes, this field is named the same way — this way the callers don't need to be changed. 

This decoder is a container of methods for decoding data in index files and allows overriding them. This change compared to the postings encoding part is a bit more involved because we the compactor might compact blocks with different postings formats in one group hence there's a need to dynamically do that.

Comment (yeya24) [reply tsdb/index/index.go]: Renamed the field `Postings` to `DecodePostings`. @bboreham

File: tsdb/compact.go
Comment (bboreham) [inline tsdb/compact.go:211]: There are three places where `nil` is reinterpreted as a default.  Do we need three?

File: tsdb/index/index.go
Comment (bboreham) [inline tsdb/index/index.go:1274]: There are three places where `nil` is reinterpreted as a default.  Do we need three?

File: tsdb/block.go
Comment (bboreham) [inline tsdb/block.go:351]: There are three places where `nil` is reinterpreted as a default.  Do we need three?

File: tsdb/compact.go
Comment (bboreham) [inline tsdb/compact.go:155]:
```suggestion
	// If it is nil then a default decoder is used, compatible with Prometheus v2.
```

I don't think "raw" is very meaningful here, not is it a a good idea to send the reader off to a different package.

**Review by bboreham — COMMENTED** [review]
Thanks, I am no longer confused.
The asymmetry between `PostingsEncoder` and `PostingsDecoderFactory` is because the latter should deal with multiple kinds of block, right?

Should `tsdb.Options` have an option to choose the encoder?  (in a different PR)

I note the new feature is not tested, beyond using the default. I'm ok with that since it's simple.

Couple of other nits below.

File: tsdb/index/index.go
Comment (yeya24) [reply tsdb/index/index.go]: Removed this one and `newReader` always requires to pass an not nill function

File: tsdb/compact.go
Comment (yeya24) [reply tsdb/compact.go]: Updated

File: tsdb/index/index.go
Comment (bboreham) [reply tsdb/index/index.go]: Thanks; we're down to two now and neither is `newReader`, so can you comment on those two please?

**Comment by bboreham** [comment]
I believe the CI failure would be fixed by #14100, if you would like to rebase.

**Comment by yeya24** [comment]
Thanks, just rebased on main.

**Comment by aknuds1** [comment]
Could you rebase again, and fix conflicts?

**Comment by yeya24** [comment]
Thanks. Rebased the PR @aknuds1

**Comment by aknuds1** [comment]
Sorry, this needs rebasing again :/

**Comment by yeya24** [comment]
@aknuds1 Rebase done.

**Review by aknuds1 — APPROVED** [review]
LGTM

**Pull request merged**
